solr/core/src/java/org/apache/solr/cluster/placement/impl/MetricAttributeImpl.java

```diff
@@ -0,0 +1,138 @@
+
+import java.util.Objects;
+import java.util.function.Function;
+
+/**
+ * Base class for {@link MetricAttribute} implementations.
+ */
+public class MetricAttributeImpl<T> implements MetricAttribute<T> {
+
+  public static final double GB = 1024 * 1024 * 1024;
+
+  /**
+   * Identity converter. It returns the raw value unchanged IFF
+   * the value's type can be cast to the generic type of this attribute,
+   * otherwise it returns null.
+   */
+  @SuppressWarnings("unchecked")
+  public final Function<Object, T> IDENTITY_CONVERTER = v -> {
+    try {
+      return (T) v;
+    } catch (ClassCastException cce) {
+      return null;
+    }
+  };
+
+  /**
+   * Bytes to gigabytes converter. Supports converting number or string
+   * representations of raw values expressed in bytes.
+   */
+  @SuppressWarnings("unchecked")
+  public static final Function<Object, Double> BYTES_TO_GB_CONVERTER = v -> {
+    double sizeInBytes;
+    if (!(v instanceof Number)) {
+      if (v == null) {
+        return null;
+      }
+      try {
+        sizeInBytes = Double.valueOf(String.valueOf(v)).doubleValue();
+      } catch (Exception nfe) {
+        return null;
+      }
+    } else {
+      sizeInBytes = ((Number) v).doubleValue();
+    }
+    return sizeInBytes / GB;
+  };
+
+  protected final String name;
+  protected final String internalName;
+  protected final Function<Object, T> converter;
+
+  /**
+   * Create a metric attribute.
+   * @param name short-hand name that identifies this attribute.
+   * @param internalName internal name of a Solr metric.
+   */
+  public MetricAttributeImpl(String name, String internalName) {
+    this(name, internalName, null);
+  }
+
+  /**
+   * Create a metric attribute.
+   * @param name short-hand name that identifies this attribute.
+   * @param internalName internal name of a Solr metric.
+   * @param converter optional raw value converter. If null then
+   *                  {@link #IDENTITY_CONVERTER} will be used.
+   */
+  public MetricAttributeImpl(String name, String internalName, Function<Object, T> converter) {
+    Objects.requireNonNull(name);
+    Objects.requireNonNull(internalName);
+    this.name = name;
+    this.internalName = internalName;
+    if (converter == null) {
+      this.converter = IDENTITY_CONVERTER;
+    } else {
+      this.converter = converter;
+    }
+  }
+
+  @Override
+  public String getName() {
+    return name;
+  }
+
+  @Override
+  public String getInternalName() {
+    return internalName;
+  }
+
+  @Override
+  public Function<Object, T> getConverter() {
+    return converter;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    MetricAttribute<?> that = (MetricAttribute<?>) o;
+    return name.equals(that.getName()) && internalName.equals(that.getInternalName()) && converter.equals(that.getConverter());
```

Comment:
Why the inconsistency between direct field access for `this` instance and getter for `that`?

Reply:
Ah, this was a problem with class hierarchy - it should cast to `MetricAttributeImpl`, I'll fix this.